lambda/matching.ml

```diff
@@ -3601,7 +3639,11 @@ let rec compile_match ~scopes repr partial ctx
 and compile_match_nonempty ~scopes repr partial ctx
     (m : (args, Typedtree.pattern Non_empty_row.t clause) pattern_matching) =
   match m with
-  | { cases = []; args = [] } -> comp_exit ctx m.default
+  | { cases = []; args = [] } ->
+      begin match comp_exit partial ctx m.default with
+      | None -> fatal_error "Matching: impossible empty matrix in a Total match"
```

Comment:
I'm not fond of leaving "fatal errors" in the compiler codebase, and even less fond of adding them. This one is in fact not new, it was present before #13076, it corresponds to the following case: https://github.com/ocaml/ocaml/blob/74a76cc2f2588eb18a443a60918f040838292ac7/lambda/matching.ml#L3310 . This means that any code that was accepted by the compiler in 5.2 will remain accepted with the proposed change.

(The new `fatal_error` appears to be raised less often than the previous one: before `comp_exit` would fail unconditionally when the default environment is empty, and now we only fail when the match is known to be Total. But note that before #13076 there was an extra catch-all clause added to the default environment for partial matches, so partial matches would never see an empty environment. In other words, the conditions to raise the fatal error are in fact the same as before.)